File: optimize-images.js
#!/usr/bin/env node
/**
 * Image optimization script for Bikurim project using Sharp
 * Compresses large product images to improve website performance.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import sharp from 'sharp';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function optimizeImage(inputPath, outputPath, maxWidth = 800, quality = 85) {
    try {
        const originalStats = fs.statSync(inputPath);
        const originalSize = originalStats.size / 1024 / 1024; // MB

        await sharp(inputPath)
            .resize(maxWidth, null, { 
                withoutEnlargement: true,
                fit: 'inside'
            })
            .jpeg({ 
                quality, 
                progressive: true,
                mozjpeg: true 
            })
            .toFile(outputPath);

        const newStats = fs.statSync(outputPath);
        const newSize = newStats.size / 1024 / 1024; // MB
        const reduction = ((originalSize - newSize) / originalSize * 100);

        console.log(`✅ ${path.basename(inputPath)}: ${originalSize.toFixed(2)}MB → ${newSize.toFixed(2)}MB (-${reduction.toFixed(1)}%)`);

    } catch (error) {
        console.error(`❌ Error optimizing ${inputPath}:`, error.message);
    }
}

async function main() {
    const imagesDir = path.join(__dirname, 'public', 'images', 'products');
    const backupDir = path.join(__dirname, 'public', 'images', 'products_backup');

    if (!fs.existsSync(imagesDir)) {
        console.error(`❌ Directory ${imagesDir} not found!`);
        return;
    }

    console.log('🔧 Optimizing product images with Sharp...');
    console.log('=' .repeat(60));

    // Create backup directory
    if (!fs.existsSync(backupDir)) {
        fs.mkdirSync(backupDir, { recursive: true });
        console.log(`📁 Created backup directory: ${backupDir}`);
    }

    const files = fs.readdirSync(imagesDir);
    const imageFiles = files.filter(file => 
        /\.(jpg|jpeg|png|webp)$/i.test(file)
    );

    if (imageFiles.length === 0) {
        console.log('❌ No image files found to optimize');
        return;
    }

    for (const filename of imageFiles) {
        const inputPath = path.join(imagesDir, filename);
        const backupPath = path.join(backupDir, filename);
        const tempPath = path.join(imagesDir, `temp_${filename}`);

        try {
            // Create backup
            fs.copyFileSync(inputPath, backupPath);

            // Optimize to temporary file
            await optimizeImage(inputPath, tempPath, 800, 85);

            // Replace original with optimized
            fs.renameSync(tempPath, inputPath);

        } catch (error) {
            console.error(`❌ Error processing ${filename}:`, error.message);
            
            // Clean up temp file if it exists
            if (fs.existsSync(tempPath)) {
                fs.unlinkSync(tempPath);
            }
        }
    }

    console.log('=' .repeat(60));
    console.log('✅ Image optimization complete!');
    console.log(`📁 Original images backed up to: ${backupDir}`);
    console.log('');
    console.log('📊 Benefits:');
    console.log('   • Faster page loading times');
    console.log('   • Reduced bandwidth usage');
    console.log('   • Better user experience');
    console.log('   • Improved mobile performance');
}

// Run the main function if this file is executed directly
main().catch(console.error);

export { optimizeImage };

File: optimize_images.py
#!/usr/bin/env python3
"""
Image optimization script for Bikurim project.
Compresses large product images to improve website performance.
"""

import os
from PIL import Image
import shutil

def optimize_image(input_path, output_path, max_width=800, quality=85):
    """
    Optimize an image by resizing and compressing it.
    
    Args:
        input_path: Path to the input image
        output_path: Path to save the optimized image
        max_width: Maximum width in pixels
        quality: JPEG quality (1-100)
    """
    try:
        # Open the image
        with Image.open(input_path) as img:
            # Convert to RGB if necessary (for JPEG)
            if img.mode in ('RGBA', 'P'):
                img = img.convert('RGB')
            
            # Calculate new dimensions
            width, height = img.size
            if width > max_width:
                ratio = max_width / width
                new_width = max_width
                new_height = int(height * ratio)
                img = img.resize((new_width, new_height), Image.LANCZOS)
            
            # Save optimized image
            img.save(output_path, 'JPEG', quality=quality, optimize=True)
            
            # Get file sizes
            original_size = os.path.getsize(input_path) / 1024 / 1024  # MB
            new_size = os.path.getsize(output_path) / 1024 / 1024  # MB
            
            print(f"✅ {os.path.basename(input_path)}: {original_size:.2f}MB → {new_size:.2f}MB ({new_size/original_size*100:.1f}%)")
            
    except Exception as e:
        print(f"❌ Error optimizing {input_path}: {e}")

def main():
    """Main function to optimize all product images."""
    images_dir = "public/images/products"
    
    if not os.path.exists(images_dir):
        print(f"❌ Directory {images_dir} not found!")
        return
    
    print("🔧 Optimizing product images...")
    print("=" * 50)
    
    # Create backup directory
    backup_dir = "public/images/products_backup"
    if not os.path.exists(backup_dir):
        os.makedirs(backup_dir)
        print(f"📁 Created backup directory: {backup_dir}")
    
    # Process each image
    for filename in os.listdir(images_dir):
        if filename.lower().endswith(('.jpg', '.jpeg', '.png')):
            input_path = os.path.join(images_dir, filename)
            backup_path = os.path.join(backup_dir, filename)
            
            # Create backup
            shutil.copy2(input_path, backup_path)
            
            # Optimize image
            optimize_image(input_path, input_path, max_width=800, quality=85)
    
    print("=" * 50)
    print("✅ Image optimization complete!")
    print(f"📁 Original images backed up to: {backup_dir}")

if __name__ == "__main__":
    main()